src/components/control/date.jsx

```diff
@@ -0,0 +1,240 @@
+  const ref = useForwardedRef(forwardedRef);
+  const [active, setActive] = useState(false);
+  const [editing, setEditing] = useState(!!props.validationMessage);
+  const [expanded, setExpanded] = useState(false);
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(fromFullDateFormat(props.value));
+  const classNames = {
+    container: undefined,
+    input: styles.input,
+    invalidInput: styles['invalid-input'],
+    calendar: styles['calendar-container'],
+    ...props.classNames,
+  };
+  const ids = {
+    input: props.id,
+    label: `${props.id}-label`,
+    tooltip: `${props.id}-tooltip`,
+    validationMessage: `${props.id}Hint`,
+  };
+  const refs = {
+    container: useRef(),
+    input: useRef(),
+  };
+
+  function onBlur(event) {
+    if (refs.container.current.contains(event.relatedTarget)) return;
+
+    // Given the validation spec, any provided validation messages are cleared on blur.
+    // If the native node is still invalid then re-render with the native validation messages.
+    refs.input.current.setCustomValidity('');
+    setValidationMessage(refs.input.current.validationMessage);
+    setActive(false);
+    if (!refs.input.current.validationMessage) setEditing(false);
+  }
+
+  function onInputChange() {
+    setValue(fromFullDateFormat(refs.input.current.value));
+  }
+
+  function onInputClick(event) {
+    if (props.readOnly) return;
+    event.preventDefault();
+    setActive(true);
+    setEditing(true);
+    setExpanded(true);
+  }
+
+  function onInputKeyDown(event) {
+    setActive(true);
+
+    if (props.readOnly) return;
+
+    if (event.key === ' ') {
+      event.preventDefault();
+      setEditing(true);
+      setExpanded(true);
+    }
+
+    if (event.key === 'Escape') {
+      setExpanded(false);
+    }
+  }
+
+  function onIconPress() {
+    setActive(true);
+    setExpanded(!expanded);
+  }
+
+  function onCalendarChange(newDate) {
+    setValue(newDate);
+    setEditing(false);
+    setExpanded(false);
+  }
+
+  useDropdownClose(refs.container, expanded, () => setExpanded(false));
+
+  useEffect(() => {
+    setEditing(!!props.validationMessage);
+    setValidationMessage(props.validationMessage);
+  }, [props.validationMessage]);
+
+  useEffect(() => {
+    setValue(fromFullDateFormat(props.value));
+  }, [props.value]);
+
+  useEffect(() => {
+    if (active && props.onChange) props.onChange({ target: ref.current });
+  }, [value]);
+
+  useLayoutEffect(() => {
+    if (active) refs.input.current.focus();
+  }, [editing]);
+
+  useLayoutEffect(() => {
+    refs.input.current.setCustomValidity(validationMessage);
+    props.onInvalid({
+      target: ref.current,
+      detail: {
+        validationMessage,
+      },
+    });
+  }, [validationMessage]);
+
+  useLayoutEffect(() => {
+    refs.input.current.value = editing ? toFullDateFormat(value) : toDateStringFormat(value);
+  }, [value]);
```

Comment:
Should this dep array also include `editing`? 🤔

Reply:
I don't think it matters here:

- If the value changes then we re-format and update the DOM reference
- If editing changes, we are already re-formatting the in-memory value and passing it down as a `value` or `defaultValue` prop IIRC

Reply:
Cool, sounds good!